packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -104,17 +117,139 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// This class stores the information of the selection under
+/// a [Selectable] or [SelectionHandler].
+///
+/// The [SelectedContentRange]s for a given [Selectable] or
+/// [SelectionHandler] can be retrieved by calling
+/// [SelectionHandler.getSelections].
+@immutable
+class SelectedContentRange {
+  /// Creates a [SelectedContentRange] with the given values.
+  const SelectedContentRange({
+    required this.contentLength,
+    required this.contentStart,
+    required this.startOffset,
+    required this.endOffset,
+  });
+
+  /// A selected content range that contains nothing.
+  static SelectedContentRange empty({int? contentLength}) {
+    return SelectedContentRange(
+      startOffset: -1,
+      endOffset: -1,
+      contentStart: -1,
+      contentLength: contentLength ?? -1,
+    );
+  }
+
+  /// The length of the content.
+  ///
+  /// The absolute value of the difference between the start
+  /// offset and end offset contained by this [SelectedContentRange]
+  /// must not exceed the content length.
+  final int contentLength;
+
+  /// The offset where the content begins.
+  final int contentStart;
+
+  /// The start of the selection relative to the start of the content.
+  ///
+  /// {@template flutter.rendering.selection.SelectedContentRange.selectionOffsets}
+  /// For example a [Text] widget's content is in the format
+  /// of an [InlineSpan] tree.
+  ///
+  /// Take the [Text] widget and [TextSpan] tree below:
+  ///
+  /// {@tool snippet}
+  /// ```dart
+  /// const Text.rich(
+  ///   TextSpan(
+  ///     text: 'Hello world, ',
+  ///     children: <InlineSpan>[
+  ///       WidgetSpan(
+  ///         child: Text('how are you today?'),
+  ///       ),
+  ///     ],
+  ///   ),
+  /// )
+  /// ```
+  /// {@end-tool}
+  ///
+  /// If we select from the beginning of 'world' to the
+  /// end of the '?' in the [WidgetSpan], we will receive
+  /// two [SelectedContentRange]s from [SelectionHandler.getSelections].
```

Comment:
At first I was confused about how this appears to be "non-flat". But it seems a SelectionListener passes you a SelectionDetails that is totally flat. So I guess these docs here are referring to a lower level part and SelectionListener is the high level interface that most users will see.

Probably nothing to really clarify about that and it's just my problem with wrapping my head around this, but I wanted to point it out in case there's any way to make that more clear.

Reply:
This Text widget now produces just one `SelectedContentRange`.